include/deal.II/lac/affine_constraints.templates.h

```diff
@@ -13,35 +13,1447 @@
+  // efficient. also, we have to do it only once, rather than in each
+  // iteration
+  size_type iteration = 0;
+  while (true)
+    {
+      bool chained_constraint_replaced = false;
+
+      for (typename std::vector<ConstraintLine>::iterator line = lines.begin();
+           line != lines.end();
+           ++line)
+        {
+#ifdef DEBUG
+          // we need to keep track of how many replacements we do in this line,
+          // because we can end up in a cycle A->B->C->A without the number of
+          // entries growing.
+          size_type n_replacements = 0;
+#endif
+
+          // loop over all entries of this line (including ones that we
+          // have appended in this go around) and see whether they are
+          // further constrained. ignore elements that we don't store on
+          // the current processor
+          size_type entry = 0;
+          while (entry < line->entries.size())
+            if (((local_lines.size() == 0) ||
+                 (local_lines.is_element(line->entries[entry].first))) &&
+                is_constrained(line->entries[entry].first))
+              {
+                // ok, this entry is further constrained:
+                chained_constraint_replaced = true;
+
+                // look up the chain of constraints for this entry
+                const size_type dof_index = line->entries[entry].first;
+                const number    weight    = line->entries[entry].second;
+
+                Assert(dof_index != line->index,
+                       ExcMessage("Cycle in constraints detected!"));
+
+                const ConstraintLine *constrained_line =
+                  &lines[lines_cache[calculate_line_index(dof_index)]];
+                Assert(constrained_line->index == dof_index,
+                       ExcInternalError());
+
+                // now we have to replace an entry by its expansion. we do
+                // that by overwriting the entry by the first entry of the
+                // expansion and adding the remaining ones to the end,
+                // where we will later process them once more
+                //
+                // we can of course only do that if the DoF that we are
+                // currently handle is constrained by a linear combination
+                // of other dofs:
+                if (constrained_line->entries.size() > 0)
+                  {
+                    for (size_type i = 0; i < constrained_line->entries.size();
+                         ++i)
+                      Assert(dof_index != constrained_line->entries[i].first,
+                             ExcMessage("Cycle in constraints detected!"));
+
+                    // replace first entry, then tack the rest to the end
+                    // of the list
+                    line->entries[entry] = std::make_pair(
+                      constrained_line->entries[0].first,
+                      constrained_line->entries[0].second * weight);
```

Comment:
@dealii/dealii This is annoying. In case this gets instantiated for `std::complex<float>` this call to `make_pair` throws a warning with gcc (clang is fine):
```
Building CXX object source/lac/CMakeFiles/obj_lac_debug.dir/affine_constraints.cc.o
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/bits/stl_algobase.h:64:0,
                 from /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/bits/stl_tree.h:63,
                 from /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/map:60,
                 from /usr/include/openmpi/ompi/mpi/cxx/mpicxx.h:40,
                 from /usr/include/x86_64-pc-linux-gnu/mpi.h:2703,
                 from /usr/include/mpi.h:10,
                 from /home/tamiko/workspace/dealii/build/include/deal.II/base/config.h:409,
                 from /home/tamiko/workspace/dealii/include/deal.II/lac/affine_constraints.h:19,
                 from /home/tamiko/workspace/dealii/source/lac/affine_constraints.cc:16:
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/bits/stl_pair.h: In instantiation of ‘constexpr std::pair<typename std::__decay_and_strip<_Tp>::__type, typename std::__decay_and_strip<_T2>::__type> std::make_pair(_T1&&, _T2&&) [with _T1 = const unsigned int&; _T2 = std::complex<float>; typename std::__decay_and_strip<_T2>::__type = std::complex<float>; typename std::__decay_and_strip<_Tp>::__type = unsigned int]’:
/home/tamiko/workspace/dealii/include/deal.II/lac/affine_constraints.templates.h:461:58:   required from ‘void dealii::AffineConstraints<number>::close() [with number = std::complex<float>]’
/home/tamiko/workspace/dealii/build/source/lac/affine_constraints.inst:22:16:   required from here
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/bits/stl_pair.h:519:5: note: the ABI of passing structure with complex float member has changed in GCC 4.4
     make_pair(_T1&& __x, _T2&& __y)
     ^~~~~~~~~
```

Reply:
Here, we could just use the explicit `std::pair<size_type, number>(...)`. That is even not much longer. When I try to compile this with `gcc-6.2.0` there is at least one more place which is not that obvious since it seems to be triggered by `boost`. Maybe, we should consider passing `-Wno-psabi `?

Reply:
That's also what I thought. Except it didn't help.

Reply:
On `gcc-6.2.0` I can avoid the warning. Do you still see it?